projects/knora-ui/src/lib/viewer/values/color-value/color-value.component.ts

```diff
@@ -0,0 +1,115 @@
+import { CreateColorValue, ReadColorValue, UpdateColorValue } from '@knora/api';
+import { Subscription } from 'rxjs';
+import { BaseValueComponent } from '../base-value.component';
+import { CustomRegex } from '../custom-regex';
+
+@Component({
+  selector: 'kui-color-value',
+  templateUrl: './color-value.component.html',
+  styleUrls: ['./color-value.component.scss']
+})
+export class ColorValueComponent extends BaseValueComponent implements OnInit, OnChanges, OnDestroy {
+
+  @Input() displayValue?: ReadColorValue;
+
+  valueFormControl: FormControl;
+  commentFormControl: FormControl;
+
+  form: FormGroup;
+
+  valueChangesSubscription: Subscription;
+
+  customValidators = [Validators.pattern(CustomRegex.COLOR_REGEX)];
+
+  constructor(@Inject(FormBuilder) private fb: FormBuilder) {
+    super();
+  }
+
+  getInitValue(): string | null {
+    if (this.displayValue !== undefined) {
+      return this.displayValue.color;
+    } else {
+      return null;
+    }
+  }
+
+  ngOnInit() {
+    // initialize form control elements
+    this.valueFormControl = new FormControl(null);
+    this.commentFormControl = new FormControl(null);
+
+    // subscribe to any change on the comment and recheck validity
+    this.valueChangesSubscription = this.commentFormControl.valueChanges.subscribe(
+      data => {
+        this.valueFormControl.updateValueAndValidity();
+      }
+    );
+
+    this.form = this.fb.group({
+      colorValue: this.valueFormControl,
+      comment: this.commentFormControl
+    });
+
+    this.resetFormControl();
+  }
+
+  ngOnChanges(changes: SimpleChanges): void {
+    this.resetFormControl();
+  }
+
+  // unsubscribe when the object is destroyed to prevent memory leaks
+  ngOnDestroy(): void {
+    this.unsubscribeFromValueChanges();
+  }
+
+  getNewValue(): CreateColorValue | false {
+    if (this.mode !== 'create' || !this.form.valid) {
+      return false;
+    }
+
+    const newColorValue = new CreateColorValue();
+
+    newColorValue.color = this.valueFormControl.value;
+
+    if (this.commentFormControl.value !== null && this.commentFormControl.value !== '') {
+      newColorValue.valueHasComment = this.commentFormControl.value;
+    }
+
+    return newColorValue;
+  }
+
+  getUpdatedValue(): UpdateColorValue | false {
+    if (this.mode !== 'update' || !this.form.valid) {
+      return false;
+    }
+
+    const updatedColorValue = new UpdateColorValue();
+
+    updatedColorValue.id = this.displayValue.id;
+
+    updatedColorValue.color = this.valueFormControl.value;
+
+    // add the submitted comment to updatedIntValue only if user has added a comment
+    if (this.commentFormControl.value !== null && this.commentFormControl.value !== '') {
+      updatedColorValue.valueHasComment = this.commentFormControl.value;
+    }
+
+    return updatedColorValue;
+  }
+
+  // update dynamically the background-color and the input label according to the picked color
+  onColorChanged(updatedValue: any) {
```

Comment:
@tobiasschweizer I have set this method, onColorChanged(), to update the label and the background-color of the input before hitting the Save button (like boolean-value component but here it is the actual value old/new state, not hard-code text true/false). However the reset does not work as expected and the override method used in boolean-value component is not transposable because I cannot catch the "old" value... Any suggestion? thanks

Reply:
You should not directly manipulate/update the `displayValue`. When the update has been successfully performed, the new value fetched from Knora will be written back.

Reply:
If you change the `displayValue`, the reset does not work properly.